### PR TITLE
Edge packages - HOLD merge

### PR DESCRIPTION
**Pull request by elvinm**
# Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Checklist:

- [x] All new and existing tests passed.
- [x] The code follows the code style and conventions of this project.
- [x] The change requires a change to the documentation.
- [x] I have updated the documentation accordingly. 

# Description of the change
Addition of a packages controller meant to support multiple types of packages. Currently the only type supported is an EdgeManifest which provides the details behind an edge deployment.

The packages controller supports:
* Post - To upload the manifest
* Get - For retrieval of the package (in current iteration the manifest)
* GetAll - To get the full list of all packages (pagination not currently supported)
* Delete - To remove a particular package given the package Id.

Example usage can be seen inside of integration tests in the pull request [here](https://github.com/Azure/azure-iot-pcs-remote-monitoring-dotnet-tests/pull/42).

# Motivation for the change
Provides the ability to upload arbitrary documents (currently limited to text format) which will be used when creating an edge deployment.

This is the java equivalent of [this pull request](https://github.com/Azure/remote-monitoring-services-dotnet/pull/103)

### CODE REVIEW

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/models/Package.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        this.id = id;
+        this.name = name;
+        this.type = type;
+        this.content = content;
+        this.dateCreated = dateCreated;
+    }
+
+    @JsonProperty("Id")
+    public String getId() {
+        return id;
```
[the remark this answers is not in the document]
Comment (elvinm) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/models/Package.java]: Done

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/services/models/Package.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    private String name;
+    private PackageType type;
+    private String content;
+    private String dateCreated;
+
+    public Package() {
+    }
+
+    public Package(String id, String name, PackageType type, String content) {
+        this(id, name, type, content,StringUtils.EMPTY);
```
[the remark this answers is not in the document]
Comment (elvinm) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/services/models/Package.java]: Done

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/webservice/v1/models/PackageListApiModel.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+import java.util.stream.StreamSupport;
+
+public class PackageListApiModel {
+
+    private Iterable<PackageApiModel> items;
+    private Hashtable<String, String> metadata;
+
+    @JsonProperty("items")
+    public Iterable<PackageApiModel> getItems() {
+        return items;
```
[the remark this answers is not in the document]
Comment (elvinm) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/webservice/v1/models/PackageListApiModel.java]: Done

File: config/app/com/microsoft/azure/iotsolutions/uiconfig/webservice/v1/models/PackageApiModel.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    public String getDateCreated() {
+        return dateCreated;
+    }
+
+    public void setDateCreated(String dateCreated) {
+        this.dateCreated = dateCreated;
+    }
+
+    public PackageApiModel(Package model) {
+        id = model.getId();
```
[the remark this answers is not in the document]
Comment (elvinm) [reply config/app/com/microsoft/azure/iotsolutions/uiconfig/webservice/v1/models/PackageApiModel.java]: Done

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/services/helpers/HttpRequestHelper.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import play.Logger;
+import play.libs.ws.WSRequest;
+import play.libs.ws.WSResponse;
+
+public class HttpRequestHelper {
+    private static final Logger.ALogger log = Logger.of(HttpRequestHelper.class);
+
+    public static void checkStatusCode(WSResponse response, WSRequest request)
+            throws ResourceNotFoundException, ConflictingResourceException, ExternalDependencyException {
+        if (response.getStatus() == 200) {
```
Comment (molly-moen) [inline iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/services/helpers/HttpRequestHelper.java]: should this check for any 2xx status code instead?

File: iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/services/models/DeploymentServiceModel.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+                                  final String packageContent,
+                                  final int priority,
+                                  final DeploymentType type) {
+        this.deviceGroupId = deviceGroupId;
+        this.deviceGroupQuery = deviceGroupQuery;
+        this.packageContent = packageContent;
+        this.name = name;
+        this.priority = priority;
+        this.type = type;
+    }
```
Comment (molly-moen) [inline iothub-manager/app/com/microsoft/azure/iotsolutions/iothubmanager/services/models/DeploymentServiceModel.java]: add blank line between methods

File: iothub-manager/test/com/microsoft/azure/iotsolutions/iothubmanager/services/DeploymentsTest.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+                 "depname, , dvcGroupQuery, true, 10, true",
+                 "depname, dvcgroupid, , true, 10, true",
+                 "depname, dvcgroupid, dvcGroupQuery, false, 10, true",
+                 "depname, dvcgroupid, dvcGroupQuery, true, -1, true"})
+    public void createDeploymentTest(String deploymentName, String deviceGroupId,
+                                    String dvcGroupQuery, boolean addPackageContent, int priority,
+                                    boolean exceptionExpected) throws Exception {
+        // Provides a different value to ensure that the configuration object returned
+        // from creating the deployment is different than the one provided to the registry manager
+        final String registryManagerDeploymentId = "test-config";
```
Comment (molly-moen) [inline iothub-manager/test/com/microsoft/azure/iotsolutions/iothubmanager/services/DeploymentsTest.java]: nit: add arrange, act, assert to these tests

**Comment by elvinm** [comment]
Closing PR as this is no longer needed as it was already merged separately.

**Pull request closed without merging**